python/lsst/daf/butler/datastores/inMemoryDatastore.py

```diff
@@ -354,28 +375,47 @@ def remove(self, ref):
         ------
         FileNotFoundError
             Attempt to remove a dataset that does not exist.
+        """
+
+        log.debug("Trash %s in datastore %s", ref, self.name)
+
+        # Check that this dataset is known to datastore
+        self._get_dataset_info(ref)
+
+        # Move datasets to trash table
+        self._move_to_trash_in_registry(ref)
+
+    def emptyTrash(self):
+        """Remove all datasets from the trash.
 
+        Notes
+        -----
+        The internal tracking of datasets is affected by this method and
+        transaction handling is not supported if there is a problem before
+        the datasets themselves are deleted.
         """
-        try:
-            storedItemInfo = self.getStoredItemInfo(ref)
-        except KeyError:
-            raise FileNotFoundError(f"No such file dataset in memory: {ref}") from None
-        thisID = ref.id
-        if storedItemInfo.parentID is not None:
-            thisID = storedItemInfo.parentID
+        log.debug("Emptying trash in datastore %s", self.name)
+        trashed = self.registry.getTrashedDatasets(self.name)
+
+        artifactsToRemove = set()
+
+        for ref in trashed:
+            realID, _ = self._get_dataset_info(ref)
+
+            allRefs = self.related[realID]
+            theseRefs = {r.id for r in itertools.chain([ref], ref.components.values())}
+            remainingRefs = allRefs - theseRefs
+            if not remainingRefs:
+                artifactsToRemove.add(realID)
 
-        if thisID not in self.datasets:
-            raise FileNotFoundError("No such file dataset in memory: {}".format(ref))
+            # Remove this entry
+            self.removeStoredItemInfo(ref)
 
-        # Only delete if this is the only dataset associated with this data
-        allRefs = self.related[thisID]
-        theseRefs = {r.id for r in itertools.chain([ref], ref.components.values())}
-        remainingRefs = allRefs - theseRefs
-        if not remainingRefs:
-            del self.datasets[thisID]
+        self.registry.emptyDatasetLocationsTrash(self.name, trashed)
 
-        # Remove rows from registries
-        self._remove_from_registry(ref)
+        for refId in artifactsToRemove:
+            log.debug("Removing artifact %s from datastore %s", realID, ref)
+            del self.datasets[refId]
```

Comment:
I think it might be safer to do this operation (unlike the others) one dataset at a time, by first deleting the artifact, then deleting the internal records, and only then removing from the trash table.  That means a Datastore would _have_ to be willing to silently ignore artifact deletion if it fails because the artifact is already gone (i.e. because a previous deletion operation only got that far before it died).  But I think that's reasonable, and the failure mode of the current logic is more dangerous: a file could be left around even though it'd already been completely removed from all records (even the trash).

Reply:
One option is to do a rename in the main trash emptying loop and then a final delete afterwards. You are right though that one reason why the file is no longer there is that some other process is emptying the trash at the same time and got to that file first (and the rename would also fail in that case as well). Multiple people emptying the trash simultaneously implies that we mostly have to suck up any errors and cross fingers. This is a problem if the file is left hanging around.

The trash emptying currently tries to delete everything. Another option is for emptyTrash to have a "delete everything mode" as now but also have a "delete only these refs and children" option (ie the refs that you only shortly before told to be trashed).

Concurrent accessing does raise the possibility that one process is emptying trash whilst another is putting the dataset that has been deleted. This will fail at the moment because the internal registry will still mention the file and the file will be on disk and since the emptyTrash is in a transaction I think that will be fine.

Reply:
As we discussed a bit on slack, I think we're pretty much forced into the "ignore errors" behavior to deal with other aspects of concurrency.

For the last point on `put` collisions with deletions, I think that's a sufficiently rare case that it's fine if the `put` just fails, because it'll do so without doing any damage to the repository (as we've already done the work to make sure `put` is atomic).